packages/core/src/render3/util/attrs_utils_with_hydration_support.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DehydratedView} from '../../hydration/interfaces';
+import {Renderer} from '../interfaces/renderer';
+import {RElement} from '../interfaces/renderer_dom';
+
+/**
+ * This represents a mapping of element tag name to an attribute name that should not be reset
+ * during hydration in case the value is the same.
+ * Values are implemented as a list in case they need to be expanded in the future to accommodate
+ * more cases.
+ */
+const hydrationProtectedElementToAttributeMap = new Map<string, string>([
+  // All these elements and their attributes force the browser to reload resources when they're set
+  // again with the same value. For example, consider an `<object>` element with its `data`
+  // attribute set to `/assets/some-file.pdf`. When the browser retrieves an HTML document from the
+  // server and finishes parsing it (after the document state is set to `complete`), it loads
+  // external resources such as images, videos, audios, etc. This includes loading
+  // `/assets/some-file.pdf`. Subsequently, when Angular begins its hydration process, it attempts
+  // to call `setAttribute` on the `<object>` element again with `setAttribute('data',
+  // '/assets/some-file.pdf')`. This action forces the browser to reload the same resources, even
+  // though they have already been loaded previously.
+  ['iframe', 'src'],
+  ['embed', 'src'],
+  ['object', 'data'],
+]);
+
+export function getHydrationProtectedAttribute(tagName: string): string | undefined {
+  return hydrationProtectedElementToAttributeMap.get(
+    // Convert to lowercase so we cover both cases when the tag name is `iframe` or `IFRAME`.
+    tagName.toLowerCase(),
+  );
+}
+
+let _setAttributeWithHydrationSupport: typeof _setAttributeWithHydrationSupportImpl = (
+  hydrationInfo: DehydratedView | null,
+  renderer: Renderer,
+  nodeIndex: number,
+  element: RElement,
+  attributeName: string,
+  attributeValue: string,
+  namespace?: string | null,
+) => {
+  renderer.setAttribute(element, attributeName, attributeValue, namespace);
+};
+
+function _setAttributeWithHydrationSupportImpl(
+  hydrationInfo: DehydratedView | null,
+  renderer: Renderer,
+  nodeIndex: number,
+  element: RElement,
+  attributeName: string,
+  attributeValue: string,
+  namespace?: string | null,
+) {
+  if (
```

Comment:
For code size optimization, we can extract `hydrationInfo?.protectedAttributes` into a separate const and use this const in the code.
```suggestion
) {
  const protectedAttributes = hydrationInfo?.protectedAttributes;
  if (
```